lib/Zonemaster/Backend/DB/SQLite.pm

```diff
@@ -164,97 +165,124 @@ sub create_new_batch_job {
 
     return $new_batch_id;
 }
 
 sub create_new_test {
     my ( $self, $domain, $test_params, $minutes_between_tests_with_same_params, $batch_id ) = @_;
-    my $result;
 
-    my $priority = 10;
-    $priority = $test_params->{priority} if (defined $test_params->{priority});
-    
-    my $queue = 0;
-    $queue = $test_params->{queue} if (defined $test_params->{queue});
+    my $dbh = $self->dbh;
 
     $test_params->{domain} = $domain;
-    my $js = JSON::PP->new;
-    $js->canonical( 1 );
+    my $js                             = JSON::PP->new->canonical;
     my $encoded_params                 = $js->encode( $test_params );
     my $test_params_deterministic_hash = md5_hex( $encoded_params );
+    my $result_id;
 
-    my $query =
-        "INSERT INTO test_results (batch_id, priority, queue, params_deterministic_hash, params) SELECT "
-      . $self->dbh->quote( $batch_id ) . ", "
-      . $self->dbh->quote( $priority ) . ", "
-      . $self->dbh->quote( $queue ) . ", "
-      . $self->dbh->quote( $test_params_deterministic_hash ) . ", "
-      . $self->dbh->quote( $encoded_params )
-      . " WHERE NOT EXISTS (SELECT * FROM test_results WHERE params_deterministic_hash='$test_params_deterministic_hash' AND creation_time > datetime('now', '-$minutes_between_tests_with_same_params minute'))";
+    my $priority = $test_params->{priority};
+    my $queue = $test_params->{queue};
 
-    my $nb_inserted = $self->dbh->do( $query );
+    my ( $recent_id, $recent_hash_id ) = $dbh->selectrow_array(
+        q[SELECT id, hash_id FROM test_results WHERE params_deterministic_hash = ? AND (CAST(strftime('%s', 'now') as integer) - CAST(strftime('%s', creation_time) as integer)) < ?],
+        undef, $test_params_deterministic_hash, 60 * $minutes_between_tests_with_same_params,
+    );
 
-    ( $result ) = $self->dbh->selectrow_array(
-        "SELECT MAX(id) AS id FROM test_results WHERE params_deterministic_hash='$test_params_deterministic_hash'" );
+    if ( $recent_id ) {
+        # A recent entry exists, so return its id
+        if ( $recent_id > $self->config->force_hash_id_use_in_API_starting_from_id() ) {
+            $result_id = $recent_hash_id;
+        }
+        else {
+            $result_id = $recent_id;
+        }
+    }
+    else {
+        $dbh->do(
+            q[INSERT INTO test_results (batch_id, priority, queue, params_deterministic_hash, params, domain, test_start_time, undelegated) VALUES (?, ?,?,?,?,?, datetime('now'),?)],
+            undef,
+            $batch_id,
+            $priority,
+            $queue,
+            $test_params_deterministic_hash,
+            $encoded_params,
+            $test_params->{domain},
+            ($test_params->{nameservers})?(1):(0),
+        );
+        
+        $dbh->do(q[UPDATE test_results SET hash_id = ? WHERE params_deterministic_hash = ?], undef, substr(md5_hex(time().rand()), 0, 16), $test_params_deterministic_hash);
```

Comment:
I don't see why this update is separated from the insert statement above. Am I missing something?

Reply:
It is separated from the insert (unlike mysql and postgresql) because SQLite doesn't have the required hashing functions to be able to initialize the hash_id column properly.